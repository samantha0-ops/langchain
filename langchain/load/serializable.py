from abc import ABC, abstractmethod
from typing import TypedDict, Literal, Any

from pydantic import BaseModel


class BaseSerialized(TypedDict):
    lc: int
    type: str
    id: list[str]


class SerializedConstructor(BaseSerialized):
    type: Literal["constructor"]
    kwargs: dict[str, Any]


class SerializedSecret(BaseSerialized):
    type: Literal["secret"]


class SerializedNotImplemented(BaseSerialized):
    type: Literal["not_implemented"]


class Serializable(BaseModel, ABC):
    @property
    def lc_namespace(self) -> list[str]:
        """
        Return the namespace of the langchain object.
        eg. ["langchain", "llms", "openai"]
        """
        return self.__class__.__module__.split(".")

    @property
    def lc_secrets(self) -> dict[str, str]:
        """
        Return a map of constructor argument names to secret ids.
        eg. {"openai_api_key": "OPENAI_API_KEY"}
        """
        return dict()

    lc_kwargs: dict[str, Any] = dict()

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.lc_kwargs = kwargs

    def to_json(self) -> SerializedConstructor:
        secrets = dict()

        # Merge the lc_secrets from every class in the MRO
        for cls in self.__class__.mro():
            # Once we get to Serializable, we're done
            if cls is Serializable:
                break

            secrets.update(super(cls, self).lc_secrets)

        return {
            "lc": 1,
            "type": "constructor",
            "id": [*self.lc_namespace, self.__class__.__name__],
            "kwargs": self.lc_kwargs
            if not secrets
            else replace_secrets(self.lc_kwargs, secrets),
        }

    def to_json_not_implemented(self) -> SerializedNotImplemented:
        return {
            "lc": 1,
            "type": "not_implemented",
            "id": [*self.lc_namespace, self.__class__.__name__],
        }


def replace_secrets(root: dict[Any, Any], secrets_map: dict[str, str]):
    result = root.copy()
    for path, secret_id in secrets_map.items():
        [*parts, last] = path.split(".")
        current = result
        for part in parts:
            if part not in current:
                break
            current[part] = current[part].copy()
            current = current[part]
        if last in current:
            current[last] = {
                "lc": 1,
                "type": "secret",
                "id": [secret_id],
            }
    return result
