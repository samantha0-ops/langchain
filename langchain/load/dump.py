import json
from typing import Any

from langchain.load.serializable import Serializable


def default(obj: Any) -> Any:
    if isinstance(obj, Serializable):
        return obj.to_json()

    return json.JSONEncoder.default(json.JSONEncoder, obj)


def dumps(obj: any, *, pretty=False):
    if pretty:
        kwargs = {"indent": 2}
    else:
        kwargs = {}
    return json.dumps(obj, default=default, **kwargs)
