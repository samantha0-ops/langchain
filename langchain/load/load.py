import json
import importlib
from typing import Any

from langchain.load.serializable import Serializable


class Reviver:
    secrets_map: dict[str, str]

    def __init__(self, secrets_map: dict[str, str] = None):
        self.secrets_map = secrets_map or dict()

    def __call__(self, value: dict[str, Any]):
        if (
            value.get("lc", None) == 1
            and value.get("type", None) == "secret"
            and value.get("id", None) is not None
        ):
            [key] = value["id"]
            if key in self.secrets_map:
                return self.secrets_map[key]
            else:
                raise KeyError(f'Missing key "{key}" in load(secrets_map)')

        if (
            value.get("lc", None) == 1
            and value.get("type", None) == "not_implemented"
            and value.get("id", None) is not None
        ):
            raise NotImplementedError(
                f"Trying to load an object that doesn't implement serialization: {value}"
            )

        if (
            value.get("lc", None) == 1
            and value.get("type", None) == "constructor"
            and value.get("id", None) is not None
        ):
            [*namespace, name] = value["id"]

            # Currently, we only support langchain imports.
            if namespace[0] != "langchain":
                raise ValueError(f"Invalid namespace: {value}")

            # The root namespace "langchain" is not a valid identifier.
            if len(namespace) == 1:
                raise ValueError(f"Invalid namespace: {value}")

            mod = importlib.import_module(".".join(namespace))
            cls = getattr(mod, name)

            # The class must be a subclass of Serializable.
            if not issubclass(cls, Serializable):
                raise ValueError(f"Invalid namespace: {value}")

            # We don't need to recurse on kwargs
            # as json.loads will do that for us.
            kwargs = value.get("kwargs", dict())
            return cls(**kwargs)

        return value


def loads(text: str, *, secrets_map: dict[str, str] = None):
    return json.loads(text, object_hook=Reviver(secrets_map))
