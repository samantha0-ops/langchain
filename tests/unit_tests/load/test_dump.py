"""Test for Serializable base class"""

from langchain.chains.llm import LLMChain
from langchain.llms.openai import OpenAI
from langchain.load.dump import dumps
from langchain.prompts.prompt import PromptTemplate


def test_serialize_openai_llm(snapshot):
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    assert dumps(llm, pretty=True) == snapshot


def test_serialize_llmchain(snapshot):
    llm = OpenAI(model="davinci", temperature=0.5, openai_api_key="hello")
    prompt = PromptTemplate.from_template("hello {name}!")
    chain = LLMChain(llm=llm, prompt=prompt)
    assert dumps(chain, pretty=True) == snapshot
