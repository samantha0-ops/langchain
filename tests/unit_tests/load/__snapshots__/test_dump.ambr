# serializer version: 1
# name: test_serialize_llmchain
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "chains",
      "llm",
      "LLMChain"
    ],
    "kwargs": {
      "llm": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "llms",
          "openai",
          "OpenAI"
        ],
        "kwargs": {
          "model": "davinci",
          "temperature": 0.5,
          "openai_api_key": {
            "lc": 1,
            "type": "secret",
            "id": [
              "OPENAI_API_KEY"
            ]
          }
        }
      },
      "prompt": {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "prompts",
          "prompt",
          "PromptTemplate"
        ],
        "kwargs": {
          "input_variables": [
            "name"
          ],
          "template": "hello {name}!"
        }
      }
    }
  }
  '''
# ---
# name: test_serialize_openai_llm
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "langchain",
      "llms",
      "openai",
      "OpenAI"
    ],
    "kwargs": {
      "model": "davinci",
      "temperature": 0.5,
      "openai_api_key": {
        "lc": 1,
        "type": "secret",
        "id": [
          "OPENAI_API_KEY"
        ]
      }
    }
  }
  '''
# ---
